### PR TITLE
Added new release workflow for release automation

### PR DESCRIPTION
**Pull request by szczepiq**
The purpose of this RB is to give you heads-up about the improvements to the release automation that I work on. The algorithm is implemented and will not change much. Documentation, edge case unit tests and fixing runnability from Gradle are pending (Travis CI will fail on this PR).

The goal of the new 'release-workflow' plugin is to able to express the release automation in the build.gradle file like that:

``` groovy
releaseWorkflow {
    step releaseNeeded
    onlyIf { releaseNeeded.needed }
    step bintrayUpload
    step pullCommits
    step configureGenericGitUser, cleanup: restoreGitUser
    step commitReleaseNotes, rollback: resetReleaseNotes
    step tagNewVersion, rollback: removeNewTag
    step bumpVersion, rollback: unbumpVersion
    step pushChanges
}
```

More over, it will be very easy to test out the release locally or even individual release steps:

```
./gradlew release -PdryRun
./gradlew pullCommits -PsingleStep
```

### CODE REVIEW

**Comment by szczepiq** [comment]
Friendly ping.

**Comment by TimvdLippe** [comment]
I have not had the time to test this, will probably have time after my exams.

**Comment by bric3** [comment]
I'm lacking enough free time to review that fast. It's next on my todo list however.

**Comment by szczepiq** [comment]
Thanks guys for putting this PR in your TODO lists!

Given that this code does not change the behavior / API of Mockito java library, I will go ahead and merge it within the next day or two. I'm very happy to get review feedback before then.

Next step, I will move the release automation into a separate repository.

File: buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowExtension.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        StepConfiguration stepConfig = new StepConfiguration(config);
+
+        if (stepConfig.isEmpty()) {
+            return; //no rollback/cleanup configured
+        }
+
+        Task rollback = stepConfig.getRollback();
+        if (rollback != null) {
+            //rollbacks only run when one of the steps fails, by default we assume they don't fail
+            if (!project.hasProperty("dryRun")) { //accommodate testing
```
Comment (bric3) [inline buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowExtension.java]: Same, I'd like some constant class / interface somewhere, instead of searching strings.

File: buildSrc/src/main/groovy/org/mockito/workflow/gradle/ReleaseWorkflow.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ *  - executed only when one the release tasks fails, otherwise they are skipped
+ *  - executed in reverse order
+ *  - even if one rollback fails other rollback tasks are still executed
+ *  - rollback of task 'foo' is _not_ executed if task 'foo' fails. It is only executed if any further release step fails.
+ *   The reason for that is that when 'foo' fails, there is nothing to rollback.
+ *
+ * Cleanup tasks:
+ *  - similar to rollback with one difference - it is executed even if the entire release is successful.
+ *   Rollback tasks are only executed when one of the release steps fails.
+ * </pre>
```
Comment (bric3) [inline buildSrc/src/main/groovy/org/mockito/workflow/gradle/ReleaseWorkflow.java]: Documentation forgot to mention that step can contain either rollback or cleanup tasks but not both.

File: buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowExtension.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                rollback.setEnabled(false);
+            }
+        } else {
+            rollback = stepConfig.getCleanup();
```
Comment (bric3) [inline buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowExtension.java]: variable should be named `postTaskAction` or something like that

File: buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowPlugin.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import org.gradle.api.tasks.TaskState;
+import org.mockito.release.exec.Exec;
+import org.mockito.release.git.Git;
+import org.mockito.release.git.GitTool;
+
+public class ReleaseWorkflowPlugin implements Plugin<Project> {
+    public void apply(final Project project) {
+        //main release task
+        project.getTasks().create("release");
+        final ReleaseWorkflowExtension ext = project.getExtensions().create("releaseWorkflow", ReleaseWorkflowExtension.class, project);
```
Comment (bric3) [inline buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowPlugin.java]: constants

File: buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowPlugin.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        //setup listener, so that the rollbacks are only executed if one of the main tasks fail
+        project.getGradle().addListener(new TaskExecutionListener() {
+            public void beforeExecute(Task task) {}
+            public void afterExecute(Task task, TaskState taskState) {
+                //when one of the main step tasks fails, enable all rollback tasks
+                if (taskState.getFailure() != null && ext.steps.contains(task)) {
+                    for (Task rollback : ext.rollbacks) {
+                        rollback.setEnabled(true);
+                    }
+                }
```
Comment (bric3) [inline buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowPlugin.java]: I think this code could be a method in `ReleaseWorkflowExtension`. It doesn't look like OO at this moment.

```java
ext.rollbackIfNeeded(task, taskState);
```


```java
class ReleaseWorkflowExtension {
  // ...

  public void rollbackIfNeeded(Task task, TaskState taskState) {
    if (taskState.getFailure() != null && steps.contains(task)) {
      for (Task rollback : rollbacks) {
        rollback.setEnabled(true);
      }
    }
  }
}
```

File: buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowExtension.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        }
+
+        StepConfiguration stepConfig = new StepConfiguration(config);
+
+        if (stepConfig.isEmpty()) {
+            return; //no rollback/cleanup configured
+        }
+
+        Task rollback = stepConfig.getRollback();
+        if (rollback != null) {
```
Comment (bric3) [inline buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowExtension.java]: I'd move this if / else logic in `StepConfiguration` since this is this class that perform the check : https://github.com/mockito/mockito/pull/724/files#diff-ffbd6e3d52cfacae9bcc7e90cbd71e72R20

File: buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/StepConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    private static void validateConfig(Map<String, Task> config) {
+        if (config.isEmpty()) {
+            return;
+        }
+        if (config.size() > 1 || !(config.containsKey("rollback") || config.containsKey("cleanup"))) {
+            throw new GradleException("Step configuration can only have one setting either 'rollback' or 'cleanup'. " +
+                    "Please refer to the documentation.");
+        }
+        String key = config.keySet().iterator().next();
+        if (!(config.get(key) instanceof Task)) {
```
Comment (bric3) [inline buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/StepConfiguration.java]: This code should use instead `entrySet`

File: buildSrc/src/test/groovy/org/mockito/cd/gradle/testing/GradleSpecification.groovy [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+     * Runs Gradle with given arguments, expects it to FAIL,
+     * prints the build.gradle file to the standard output if the test fails
+     */
+    BuildResult fail(String... args) {
+        try {
+            GradleRunner.create()
+                    .withProjectDir(projectDir.root)
+                    .withArguments(args)
+                    .buildAndFail()
+        } catch (Exception e) {
```
Comment (bric3) [inline buildSrc/src/test/groovy/org/mockito/cd/gradle/testing/GradleSpecification.groovy]: miss aligned see [contributing](https://github.com/mockito/mockito/blob/release/2.x/.github/CONTRIBUTING.md)

File: gradle/release.gradle [hunk 9]
```diff
@@ -43,14 +48,9 @@ notes {
...
-if (dryRun) {
-    //TODO SF lets not use excludedTaskNames to model this
-    logger.lifecycle "Automatically scheduling 'rollbackRelease' task"
-    //mustRun instead of dependsOn to avoid running 'release' task every time use 'dryRun' property
-    rollbackRelease.mustRunAfter release
-    gradle.startParameter.taskNames += "rollbackRelease"
-}
+//dryRun project property is interpreted by the release-workflow plugin to enable rollbacks
+//releaseTest does not publish or push but it leaves the commits so that they can be inspected manually
+def releaseTest = project.hasProperty('releaseTest') || project.hasProperty('dryRun')
```
Comment (bric3) [inline gradle/release.gradle]: Maybe rename to `releaseWorkflowTest` otherwise it maybe understood as run test against release (for users that don't know much about gradle but that have shell completion)

File: gradle/release.gradle [hunk 10]
```diff
@@ -106,121 +105,147 @@ task("releaseNeeded") {
...
+        This introduces a problem - someone might have pushed changes *after* release process has started
+        What can happen:
+         - bintrayUpload will fail saying that the version is already released
+         - git push fails saying that the pull is needed
+        Resolutions:
+         - see 'TROUBLESHOOTING' section at the top of this file
+    */
+    step checkOutCommittableBranch
+    step updateReleaseNotes
+    step commitReleaseNotes, [rollback: resetReleaseNotes]
```
Comment (bric3) [inline gradle/release.gradle]: shouldn't it be before release upload ?

File: gradle/release.gradle [hunk 11]
```diff
@@ -106,121 +105,147 @@ task("releaseNeeded") {
...
-    run "git", "tag", "-a", tag, "-m", "$tagMessage" as String
+task pushChanges(type: Exec) {
+    description = "Pushes changes to '${System.env.TRAVIS_BRANCH}' branch."
+    //!!!Below command _MUST_ be quiet otherwise it exposes GitHub write token!!!
+    def mustBeQuiet = "-q"
+    def pushCommandLine = ["git", "push", "https://szczepiq:${System.env.GH_TOKEN}@github.com/mockito/mockito.git", System.env.TRAVIS_BRANCH, "v$currentVersion", mustBeQuiet]
+    if (releaseTest) {
+        pushCommandLine << '--dry-run'
+    }
+    commandLine pushCommandLine
```
Comment (bric3) [inline gradle/release.gradle]: If that fails the script can still try a `git pull --rebase`, then if successful retry the command line (`git push ...`), could be improved later though.

File: gradle/release.gradle [hunk 12]
```diff
@@ -106,121 +105,147 @@ task("releaseNeeded") {
...
- */
-
-private void commitReleaseNotes(String buildInfo) {
-    def notesFile = project.file("doc/release-notes/official.md")
-    run "git", "add", "$notesFile" as String
-    run "git", "commit", "-m", "Update release notes $buildInfo" as String, "$notesFile" as String
+task unbumpVersion(type: Exec) {
+    description = "Unbumps version in 'version.properties' by resetting head by one commit (--hard)."
+    commandLine "git", "reset", "--hard", "HEAD~1"
 }
```
Comment (bric3) [inline gradle/release.gradle]: Isn't it possible to have _`RollbackableTasks`_ where it is possible to declare the task and the rollback/cleanup ?

Same idea for dryrun Task

**Comment by szczepiq** [comment]
Thank you for thorough feedback! I'll update the code.

File: buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowExtension.java (hunk 6 above)
Comment (szczepiq) [inline buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowExtension.java]: It's not a check - rollback is nullable

File: buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowExtension.java (hunk 3 above)
Comment (szczepiq) [inline buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowExtension.java]: I really like the new name!

File: buildSrc/src/main/groovy/org/mockito/workflow/gradle/ReleaseWorkflow.java (hunk 2 above)
Comment (szczepiq) [inline buildSrc/src/main/groovy/org/mockito/workflow/gradle/ReleaseWorkflow.java]: Great catch!

File: buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowPlugin.java (hunk 5 above)
Comment (szczepiq) [inline buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowPlugin.java]: You are right, it does not look nice at the moment given all Java garbage. Let's not put it on the extension type, because that type is exposed to the user (public API) to be used in build.gradle files.

Comment (szczepiq) [inline buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowPlugin.java]: Instead, I'll roll out a new inner class.

File: buildSrc/src/test/groovy/org/mockito/cd/gradle/testing/GradleSpecification.groovy (hunk 8 above)
Comment (szczepiq) [inline buildSrc/src/test/groovy/org/mockito/cd/gradle/testing/GradleSpecification.groovy]: I fixed it. Fixing / reporting this does not seem like a valueable use of my and your time, though.

File: gradle/release.gradle (hunk 9 above)
Comment (szczepiq) [inline gradle/release.gradle]: Yes, I'm with you that the name is not great. I was trying to keep the balance between brevity and clarity. I don't see big value changing to releaseWorkflowTest.

File: gradle/release.gradle (hunk 12 above)
Comment (szczepiq) [inline gradle/release.gradle]: Yes, it should be possible. We can figure out a different api for declaring release in build.gradle. What I suggest in this PR is a good starting point.

File: gradle/release.gradle (hunk 11 above)
Comment (szczepiq) [inline gradle/release.gradle]: Yeah, interesting! So far push has not failed. If it does I'm not sure retrying would help.

File: gradle/release.gradle (hunk 10 above)
Comment (szczepiq) [inline gradle/release.gradle]: Yes, we could do that. We could make the upload to be the very last step. I think this is a very good idea. I implement it outside of this PR. This PR only refactors existing behavior. I'll add a TODO.

File: buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowExtension.java [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    private void addStep(final Task task, Map<String, Task> config) {
+        //populate steps collection
+        steps.add(task);
+
+        //main release task will depend on this step
+        project.getTasks().getByName("release").dependsOn(task);
+
+        //release steps must be sequential
+        if (previousStep != null) {
+            if (project.hasProperty("singleStep")) {
```
Comment (szczepiq) [inline buildSrc/src/main/groovy/org/mockito/workflow/gradle/internal/ReleaseWorkflowExtension.java]: Constants in such cases are the matter of personal style. I find String literal like that more readable. I will skip extracting constants at this time. However, I don't mind if you or somebody else refactors it when working on this code.

**Comment by szczepiq** [comment]
Sweet review, thank you. It's really cool that you're getting familiar with the release logic!!!

File: gradle/release.gradle (hunk 12 above)
Comment (bric3) [inline gradle/release.gradle]: Yes it was just a comment.

File: gradle/release.gradle (hunk 10 above)
Comment (bric3) [inline gradle/release.gradle]: yes sure

**Comment by bric3** [comment]
@szczepiq You should rather perform a `git rebase release/2.x` instead of merging release/2.x in the branch, this keep history a bit cleaner in branches.

**Comment by szczepiq** [comment]
Thanks for the tip @bric3! I assumed that rebase would be tricky given that I have already pushed the branch to GitHub (e.g. changing history of already-pushed code).

Next time, I will do the rebase.

**Comment by TimvdLippe** [comment]
@szczepiq already pushing does not limit you from rebasing. It results in a lot cleaner PRs

**Comment by bric3** [comment]
Yep, github handles flawlessly push forced comment on a branch.
